### PR TITLE
Distributor might start serving writes with an empty HA tracke…

### PR DESCRIPTION
**Pull request by TimKotowski**
#### What this PR does
The distributor doesn't wait for the HA tracker to complete initializing before starting to serve writes. This means that the distributor can start serving writes with an empty HA tracker state. An empty state will lead to accepting any sample from any HA replica. Allow fix to preload our cluster cache before writing to replicas. Want to ensure we are writing to the leader replica only.
#### Which issue(s) this PR fixes or relates to

Closes https://github.com/grafana/mimir/issues/5796

#### Checklist

- [x] Tests updated.
- NA Documentation Written
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- NA [about-versioning.md](https://github.com/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/distributor/ha_tracker.go [hunk 1]
```diff
@@ -274,6 +283,41 @@ const (
...
+				delete(h.clusters, user)
+			}
+		}
+		return true
+	}
+
+	// Store the received information into our cache
+	h.electedLock.Lock()
+	h.updateCache(user, cluster, replica)
+	h.electedLock.Unlock()
```
Comment (TimKotowski) [inline pkg/distributor/ha_tracker.go:316]: I have a question, just for knowledge learning. for the `WatchPrefix` 
Comments say 
```
 Request callbacks from KVStore when data changes
```
But i'm curious how important this is as it seems to be only called once instead of staying open for any changes for N amount of time time we do for ticker's for `updateKVLoop`.

I would of thought 
`WatchPrefix` will stay open as long as this code below is still going.
```
go func() {
		defer wg.Done()
		h.updateKVLoop(ctx)
	}()
```
 Since this will do a ticker and update the KVStore or clean up Replicas. And I would of thought `WatchPrefix` would watch for those changes each time?

File: pkg/distributor/ha_tracker.go [hunk 2]
```diff
@@ -226,39 +267,11 @@ func (h *haTracker) loop(ctx context.Context) error {
...
-				delete(userClusters, cluster)
-				if len(userClusters) == 0 {
-					delete(h.clusters, user)
-				}
-			}
-			return true
+		replica, ok := value.(*ReplicaDesc)
+		if !ok {
+			return false
 		}
```
Comment (TimKotowski) [inline pkg/distributor/ha_tracker.go:273]: This was added more so as a sanity check to make sure if its not the correct type to return false. Do we agree this is something we should keep in `WatchPrefix`? For HA Tracker.

**Comment by TimKotowski** [comment]
@dimitarvdimitrov Is it okay if I get a look at this?

**Comment by dimitarvdimitrov** [comment]
is someone more familiar with the distributor up to reviewing this, perhaps @duricanikolic @ying-jeanne @aknuds1 @fayzal-g?  Otherwise, I can take a look in the first week of January

File: pkg/distributor/ha_tracker.go [hunk 3]
```diff
@@ -202,10 +202,55 @@ func newHATracker(cfg HATrackerConfig, limits haTrackerLimits, reg prometheus.Re
...
+		return err
+	}
+
+	if len(keys) == 0 {
+		level.Warn(h.logger).Log("msg", "syncHAStateOnStart: no keys for HA tracker prefix", "err", err)
+		return nil
+	}
+
+	// TODO what is the benefit to allowing an a concurrent/parallel process here
+	// What is the amount fo keys we are getting back? Is it inefficient to serially allow this behavior?
```
Comment (TimKotowski) [inline pkg/distributor/ha_tracker.go:228]: I left a TODO here, what is the benefit of allowing asynchronous behavior for getting all the replicas based on the keys coming back from the ` h.client.List`. I'm still very new to contributing, and I couldn't really find what the the average length of keys that come back from `h.client.List` is?. I dont wanna create goroutines here if its not gonna bring much benefit at all. But I would like a discussion on the amount of keys getting returned and the performance impact this could bring if left serially. Since this is a preLoad like state.

Comment (TimKotowski) [reply pkg/distributor/ha_tracker.go:228]: @dimitarvdimitrov @duricanikolic @ying-jeanne @aknuds1 @fayzal-g pinging people that seem to know a lot of distributor.

**Comment by TimKotowski** [comment]
Anyone back from holidays able to look through PR and comments?

**Comment by dimitarvdimitrov** [comment]
> Anyone back from holidays able to look through PR and comments?

sorry for the delay, I will take a look at this PR in the coming days

File: pkg/distributor/ha_tracker.go [hunk 4]
```diff
@@ -202,10 +202,55 @@ func newHATracker(cfg HATrackerConfig, limits haTrackerLimits, reg prometheus.Re
...
+	for i := 0; i < len(keys); i++ {
+		if ctx.Err() != nil {
+			return ctx.Err()
```
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker.go:231]: we've had some issues with improper context cancellation and vague context cancellation messages. Let's use `context.Cause(ctx)` instead of `ctx.Err`, and also wrap the error into something saying when the error happened: `fmt.Errorf("syncing HA tracker state on startup: %w")`

File: pkg/distributor/ha_tracker.go [hunk 5]
```diff
@@ -202,10 +202,55 @@ func newHATracker(cfg HATrackerConfig, limits haTrackerLimits, reg prometheus.Re
...
+		}
+
+		desc, ok := val.(*ReplicaDesc)
+		if !ok {
+			level.Error(h.logger).Log("msg", "syncHAStateOnStart: got invalid replica descriptor", "key", keys[i])
+			continue
+		}
+
+		if notProcessed := h.processKVStoreEntry(keys[i], desc); !notProcessed {
+			level.Warn(h.logger).Log("msg", "syncHAStateOnStart: failed to processed replica value to cache", "key", keys[i], "err", err)
```
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker.go:247]:
```suggestion
			level.Warn(h.logger).Log("msg", "sync HA state on start: failed to process replica value to cache", "key", keys[i], "err", err)
```

since those would be read by people, it's easier to understand if it's in plain English. Same applies to the other few log lines.

File: pkg/distributor/ha_tracker.go (hunk 3 above)
Comment (dimitarvdimitrov) [reply pkg/distributor/ha_tracker.go:228]: I checked a few clusters internally and etcd GET latency averages at below 5ms. It will add some latency to startup, but I don't think 5-10 seconds are crucial. If they end up causing problems for someone, we can improve it. I think your current approach is good.

File: pkg/distributor/ha_tracker_test.go [hunk 6]
```diff
@@ -792,6 +791,58 @@ func TestCheckReplicaCleanup(t *testing.T) {
...
+			EnableHATracker:        true,
+			KVStore:                kv.Config{Mock: mock},
+			UpdateTimeout:          15 * time.Second,
+			UpdateTimeoutJitterMax: 0,
+			FailoverTimeout:        time.Millisecond * 2,
+		}, trackerLimits{maxClusters: 100}, nil, log.NewNopLogger())
+
+		require.NoError(t, err)
+		require.NoError(t, services.StartAndAwaitRunning(context.Background(), c))
+		defer services.StopAndAwaitTerminated(context.Background(), c) //nolint:errcheck
```
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker_test.go:837]: let's use `t.Cleanup` instead of `defer` like in the consul instance above

File: pkg/distributor/ha_tracker.go [hunk 7]
```diff
@@ -274,6 +291,41 @@ const (
...
+	}
+
+	// Store the received information into our cache
+	h.electedLock.Lock()
+	h.updateCache(user, cluster, replica)
+	h.electedLock.Unlock()
+	h.electedReplicaPropagationTime.Observe(time.Since(timestamp.Time(replica.ReceivedAt)).Seconds())
+
+	return true
+}
```
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker.go:328]:
```suggestion
}

```

File: pkg/distributor/ha_tracker_test.go [hunk 8]
```diff
@@ -792,6 +791,58 @@ func TestCheckReplicaCleanup(t *testing.T) {
...
+		}, trackerLimits{maxClusters: 100}, nil, log.NewNopLogger())
+		require.NoError(t, err)
+		require.NoError(t, services.StartAndAwaitRunning(context.Background(), c))
+
+		now = time.Now()
+		err = c.checkReplica(context.Background(), "user", cluster, replica, now)
+		assert.NoError(t, err)
+
+		// Check to see if the value in the trackers cache is correct.
+		checkReplicaTimestamp(t, time.Second, c, "user", cluster, replica, now)
```
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker_test.go:823]: can the `"user"` also go in a variable (preferably a constant) at the top of the test?

File: pkg/distributor/ha_tracker_test.go [hunk 9]
```diff
@@ -792,6 +791,58 @@ func TestCheckReplicaCleanup(t *testing.T) {
 	))
 }
 
+// TestStartingSync tests that values are synced to KVStore on startup
+func TestStartingSync(t *testing.T) {
+	t.Run("check that if a second replica is set up during preLoad that it does not write to it since its not the leader", func(t *testing.T) {
```
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker_test.go:796]: this test description looks out of place. Since there's only one test case here can we remove the t.Run?

**Review by dimitarvdimitrov — COMMENTED** [review]
Thank you for your work. I'm starting to doubt whether this bug was ever present or if I misread the code initially.

I tried commenting out the initial sync that you added and expected that `TestStartingSync` will fail. I suspect it didn't because if the HA tracker sees a new replica which it hasn't seen before it will update the KV store before returning the value. The update happens by a compare-and-swap, and prevents from overwriting the correct leader (r1 in the test case). https://github.com/grafana/mimir/blob/06ba08b733deeb41f329f72bfa769d86dbad8978/pkg/distributor/ha_tracker.go#L530-L537

Can you verify this too?

This will be a problem if we decide to support memberlist (#1597) where compare-and-swap always returns immediately, but it's not a problem right now with etcd and consul. Because of this I still think it's worth it to merge this change.

File: pkg/distributor/ha_tracker.go [hunk 10]
```diff
@@ -202,10 +202,55 @@ func newHATracker(cfg HATrackerConfig, limits haTrackerLimits, reg prometheus.Re
...
+	// What is the amount fo keys we are getting back? Is it inefficient to serially allow this behavior?
+	for i := 0; i < len(keys); i++ {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		val, err := h.client.Get(ctx, keys[i])
+		if err != nil {
+			level.Warn(h.logger).Log("msg", "syncHAStateOnStart: failed to get replica value", "key", keys[i], "err", err)
+			return err
```
[the remark this answers is not in the document]
Comment (TimKotowski) [reply pkg/distributor/ha_tracker.go:237]: There were certain errors  I thought that matters to return out. But you are right a lot of there i should continue just logging if there is an error. The only error that makes sense to return out would be the one where if the Context was canceled so something bigger happened in the system so we should return out there, that sound okay?

File: pkg/distributor/ha_tracker.go [hunk 11]
```diff
@@ -274,6 +291,41 @@ const (
 	deletionTimeout = 30 * time.Minute
 )
 
+func (h *haTracker) processKVStoreEntry(key string, replica *ReplicaDesc) bool {
```
[the remark this answers is not in the document]
Comment (TimKotowski) [reply pkg/distributor/ha_tracker.go:294]: yeah I can. And you are 100% correct I was thinking about that. The only issue i had with keeping it is `processKVStoreEntry` is used in the `WatchPrefix`. And `processKVStoreEntry` returns out of `WatchPrefix` with that function call.

File: pkg/distributor/ha_tracker.go [hunk 12]
```diff
@@ -202,10 +202,55 @@ func newHATracker(cfg HATrackerConfig, limits haTrackerLimits, reg prometheus.Re
...
+// Will update the internal cache to be in sync with the KV Store
+func (h *haTracker) syncHAStateOnStart(ctx context.Context) error {
+	if !h.cfg.EnableHATracker {
+		return nil
+	}
+
+	// haTracker holds HATrackerConfig with the prefix.
+	keys, err := h.client.List(ctx, "")
+	if err != nil {
+		level.Warn(h.logger).Log("msg", "syncHAStateOnStart: failed to list replica keys", "err", err)
```
[the remark this answers is not in the document]
Comment (TimKotowski) [reply pkg/distributor/ha_tracker.go:218]: sounds good, ill remove.

**Comment by TimKotowski** [comment]
Thank you for your work. I'm starting to doubt whether this bug was ever present or if I misread the code initially.

I tried commenting out the initial sync that you added and expected that TestStartingSync will fail. I suspect it didn't because if the HA tracker sees a new replica which it hasn't seen before it will update the KV store before returning the value. The update happens by a compare-and-swap, and prevents from overwriting the correct leader (r1 in the test case). 

Yeah when I was writing this test thats what i thought to.

The KV store will update with the new replica as the leader right? I was a little confused to. I thought it wouldn't update with the new leader and still keep the old one. But this does not seem the case? At least for our in memory cache it will update storing the "new" leader right?

File: pkg/distributor/ha_tracker_test.go (hunk 6 above)
Comment (TimKotowski) [reply pkg/distributor/ha_tracker_test.go:837]: ill make that change. ty.

File: pkg/distributor/ha_tracker.go (hunk 3 above)
Comment (TimKotowski) [reply pkg/distributor/ha_tracker.go:228]: sounds good, thank you for this.

File: pkg/distributor/ha_tracker_test.go (hunk 9 above)
Comment (TimKotowski) [reply pkg/distributor/ha_tracker_test.go:796]: yeah i can.

File: pkg/distributor/ha_tracker.go [hunk 13]
```diff
@@ -274,6 +291,41 @@ const (
...
+		h.electedLock.Lock()
+		defer h.electedLock.Unlock()
+		userClusters := h.clusters[user]
+		if userClusters != nil {
+			delete(userClusters, cluster)
+			if len(userClusters) == 0 {
+				delete(h.clusters, user)
+			}
+		}
+		return true
```
[the remark this answers is not in the document]
Comment (TimKotowski) [reply pkg/distributor/ha_tracker.go:318]: sure thing.

**Comment by dimitarvdimitrov** [comment]
So in the context of the changes before your PR (i.e. in the context of what's on `main`:

> Based on my finding in the test, it will not update the KV Store for replica 2 aka r2 if it has not seen it before. Instead it hits here.
> 
> ```
> else {
> 			// Sample received is from non-elected replica: record details and reject.
> 			entry.nonElectedLastSeenReplica = replica
> 			entry.nonElectedLastSeenTimestamp = timestamp.FromTime(now)
> 			err = newReplicasDidNotMatchError(replica, entry.elected.Replica)
> 		}
> ```

you're right and I think this behaviour is the correct one. However, r2 doesn't end up there because of the initial syncs. When attempting with r2, the code should hit here because the in-memory cache doesn't contain the requested cluster 
https://github.com/grafana/mimir/blob/59fe89a160f5eff2fbc047fda092ebb1551b39ae/pkg/distributor/ha_tracker.go#L432-L446
then

https://github.com/grafana/mimir/blob/59fe89a160f5eff2fbc047fda092ebb1551b39ae/pkg/distributor/ha_tracker.go#L475-L493

this line in particular

https://github.com/grafana/mimir/blob/59fe89a160f5eff2fbc047fda092ebb1551b39ae/pkg/distributor/ha_tracker.go#L480-L482

so in essence the in-memory cache is lazily updated as requests come in. But an empty cache doesn't mean that any replica will be elected by the distributor.


---

I found this out with a debugger. It makes stepping through the code easier

File: pkg/distributor/ha_tracker.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/distributor/ha_tracker.go:294]: yes, in this case we can hardcode the `return true` in `WatchPrefix`.

File: pkg/distributor/ha_tracker.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/distributor/ha_tracker.go:237]: > 1. What if for the `h.client` there were bigger issue's (like a connectivity issues) which made me think the `startingFn` behavior would have issues since it uses the `h.client`

We could potentially use retries with backoff in case the KV store has issues. 

I just checked the regular HA tracker logic and changed my mind. On the regular push path we return a 5xx error in case the KV store is unavailable. I think on startup it makes more sense to follow the same logic. That is - instead of `continue`ing we return an error. We should still stick to the error tolerance that the regular checkSample has (e.g. if the key doesn't follow the expected 2-segmented format).

WDYT?

File: pkg/distributor/ha_tracker_test.go (hunk 8 above)
Comment (TimKotowski) [reply pkg/distributor/ha_tracker_test.go:823]: yeah, good catch.

File: pkg/distributor/ha_tracker.go (hunk 4 above)
Comment (TimKotowski) [reply pkg/distributor/ha_tracker.go:231]: ahh gotcha, makes sense. I'll make that change, ty.

File: pkg/distributor/ha_tracker.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (TimKotowski) [reply pkg/distributor/ha_tracker.go:294]: gotcha, sounds good, ill do that. ty.

**Comment by dimitarvdimitrov** [comment]
> Technically replica 2 is part of the same cluster as replica 1 right? so shouldn't it not hit the code flow you pointed out?

after the restart the in-memory cache is empty (in the state of the code on the `main` branch). So this forces the CAS update in the KV store. The update ends up being just a fetch and doesn't update anything in the KV store.

>Or is that what we want the change, even if its part the same cluster we should still up the KV store with the other replica's for that cluster?

I don't think so. This PR should only sync the cluster on startup, but not really skip the cache if the cache is present.

**Comment by TimKotowski** [comment]
@dimitarvdimitrov I tried commenting out the initial sync that you added and expected that TestStartingSync will fail. I suspect it didn't because if the HA tracker sees a new replica which it hasn't seen before it will update the KV store before returning the value. The update happens by a compare-and-swap, and prevents from overwriting the correct leader (r1 in the test case). 

___ From comment above

Looking through debugger as the CAS portion, you are right. CAS will happen, and it will see a new replica it hasn't seen before and update the KV store.

**Comment by dimitarvdimitrov** [comment]
> I guess this brings me to a point where should this ticket be then?

This will be a problem if we decide to support memberlist (https://github.com/grafana/mimir/issues/1597) where compare-and-swap always returns immediately, but it's not a problem right now with etcd and consul. Because of this I still think it's worth it to merge this change.

**Comment by TimKotowski** [comment]
@dimitarvdimitrov is the `memberlist` referring to https://github.com/hashicorp/memberlist?

**Comment by dimitarvdimitrov** [comment]
yes

**Comment by duricanikolic** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

**Comment by TimKotowski** [comment]
@dimitarvdimitrov I havn't worked on this because seems i saw some pull requests regarding member list, and i don't see much value in the PR, thoughts?

**Comment by dimitarvdimitrov** [comment]
I think this work will be needed for the work @aldernero is doing around memberlist for the HA tracker. Maybe he can pick this up or you two can work together.

**Comment by TimKotowski** [comment]
> I think this work will be needed for the work @aldernero is doing around memberlist for the HA tracker. Maybe he can pick this up or you two can work together.

i wouldn't mind pairing up.

**Comment by dimitarvdimitrov** [comment]
https://github.com/grafana/mimir/issues/5796#issuecomment-2473327258
